pkg/controller/syncclusterdeployment/syncclusterdeployment_controller.go

```diff
@@ -0,0 +1,445 @@
+		}
+
+	}
+
+	// find items that need deleting
+	for _, rMS := range remoteMachineSets.Items {
+		found := false
+		for _, ms := range clusterDeploymentComputeMachineSets {
+			if rMS.Name == ms.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			machineSetsToDelete = append(machineSetsToDelete, &rMS)
+		}
+	}
+
+	c.logger.Debugf("Creating machineset(s) %+v", machineSetsToCreate)
+	c.logger.Debugf("Updating machineset(s) %+v", machineSetsToUpdate)
+	c.logger.Debugf("Deleting machineset(s) %+v", machineSetsToDelete)
+
+	for _, ms := range machineSetsToCreate {
+		c.logger.Debugf("creating ms %v", ms)
+		_, err = remoteClusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterAPINamespace).Create(ms)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, ms := range machineSetsToUpdate {
+		c.logger.Debugf("updating ms %v", ms)
+		_, err = remoteClusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterAPINamespace).Update(ms)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, ms := range machineSetsToDelete {
+		c.logger.Debugf("deleting ms %v", ms)
+		err = remoteClusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterAPINamespace).Delete(ms.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// return list of non-master machinesets from a given ClusterDeployment
+func computeClusterAPIMachineSetsFromClusterDeployment(cluster *cov1.ClusterDeployment, coClusterVersion *cov1.ClusterVersion) []*clusterapiv1.MachineSet {
+	machineSets := []*clusterapiv1.MachineSet{}
+
+	for _, ms := range cluster.Spec.MachineSets {
+		if ms.NodeType == cov1.NodeTypeMaster {
+			continue
+		}
+		newMS, _ := buildClusterAPIMachineSet(&ms, coClusterVersion)
+		machineSets = append(machineSets, newMS)
+	}
+	return machineSets
+}
+
+func buildClusterAPIMachineSet(ms *cov1.ClusterMachineSet, clusterVersion *cov1.ClusterVersion) (*clusterapiv1.MachineSet, error) {
+	capiMachineSet := clusterapiv1.MachineSet{}
+	capiMachineSet.Name = ms.ShortName
+	capiMachineSet.Namespace = remoteClusterAPINamespace
+	replicas := int32(ms.Size)
+	capiMachineSet.Spec.Replicas = &replicas
+	capiMachineSet.Spec.Selector.MatchLabels = map[string]string{"machineset": ms.ShortName}
+
+	/*
+		// TODO: figure out what to do about annotations
+		if machineSet.Annotations == nil {
+			machineSet.Annotations = map[string]string{}
+		}
+		sClusterVersion, err := serializeCOResource(clusterVersion)
+		if err != nil {
+			return nil, err
+		}
+		machineSet.Annotations["cluster-operator.openshift.io/cluster-version"] = string(sClusterVersion)
+	*/
+
+	machineTemplate := clusterapiv1.MachineTemplateSpec{}
+	machineTemplate.Labels = map[string]string{"machineset": ms.ShortName}
+	machineTemplate.Spec.Labels = map[string]string{"machineset": ms.ShortName}
+
+	capiMachineSet.Spec.Template = machineTemplate
+
+	return &capiMachineSet, nil
+}
+
+func buildClusterAPICluster(cluster *cov1.ClusterDeployment) (*clusterapiv1.Cluster, error) {
```

Comment:
even better, I think we should factor out the buildMachineSet/buildCluster functions from https://github.com/openshift/cluster-operator/pull/241 into controller_utils and then use them there and here.